kqueen/engines/openstack_kubespray.py

```diff
@@ -0,0 +1,434 @@
+
+    def __init__(self, connection):
+        """Initialize Heat client wrapper.
+
+        :param openstack.connection.Connection connection: openstacksdk
+            connection instance
+        """
+        self.connection = connection
+
+    def create(self, *, stack_name, **kwargs):
+        """Create Heat stack with given kwargs."""
+
+        template_file, files = get_template_and_files()
+        try:
+            return self.connection.create_stack(
+                name=stack_name,
+                template_file=template_file,
+                files=files,
+                rollback=False,
+                **kwargs,
+            )
+        except openstack.exceptions.BadRequestException as e:
+            # NOTE(sskripnick)
+            # handle this exception separately because useful information
+            # may be found only in e.details
+            logger.exception("Failed to create stack: %s" % e.details)
+            raise
+        except Exception as e:
+            logger.exception("Failed to create stack: %s" % e)
+            raise
+
+    def get(self, stack_name):
+        return self.connection.get_stack(stack_name)
+
+    def get_outputs(self, stack):
+        outputs = {}
+        for output in stack.outputs:
+            outputs[output.output_key] = output.output_value
+        return outputs
+
+    def delete(self, name):
+        """Delete stack.
+
+        :returns: True if delete succeeded, False if the stack was not found.
+        """
+        return self.connection.delete_stack(name)
+
+
+class OpenstackKubesprayEngine(BaseEngine):
+    """OpenStack Kubespray engine.
+
+    This engine can provision k8s cluster in OpenStack cloud.
+
+    Ansible and Kubespray should be installed and configured.
+
+    Path to kubespray may be configured by setting KUBESPRAY_FILES_PATH,
+    e.g. by setting environment variable KQUEEN_KUBESPRAY_FILES_PATH.
+
+    Known to work with kubespray-2.5.0, ansible-2.4.3.0 and
+    ubuntu 16.04 image.
+
+    """
+    name = "openstack_kubespray"
+    verbose_name = "Openstack Kubespray Engine"
+    parameter_schema = {
+        "cluster": {
+            "ssh_key_name": {
+                "type": "text",
+                "order": 10,
+                "label": "SSH key name",
+                "validators": {
+                    "required": True,
+                },
+            },
+            "ssh_username": {
+                "type": "text",
+                "order": 15,
+                "label": "SSH username",
+                "default": "ubuntu",
+                "validators": {
+                    "required": True,
+                },
+            },
+            "image_name": {
+                "type": "text",
+                "order": 20,
+                "label": "Image name",
+                "validators": {
+                    "required": True,
+                },
+            },
+            "flavor": {
+                "type": "text",
```

Comment:
isn't flavors are already predefied? can we get them and show as a select field? (may be in a separate commit)

Reply:
Yes, it is nice to have.